### PR TITLE
The promise of a future

### LINKED ISSUE #4: tracking issue codegen
- [x] object
~- [ ] callback~
- [x] future
- [x] stream
- [x] option
- [x] result
- [ ] struct
- [ ] enum

### PR DESCRIPTION
**Pull request by dvc94ch**
Kind of messy, but the test passes using dart. Will add js support tomorrow.

- [x] dart finalizers regression
- [x] cleanup

Closes #4

### CODE REVIEW

File: src/rust.rs [hunk 1]
```diff
@@ -40,11 +53,82 @@ impl RustGenerator {
...
+                    #[cfg(target_family = "wasm")]
+                    let waker = waker_fn(move || {
+                        unsafe { __notifier_callback(port as _) };
+                    });
+                    #[cfg(not(target_family = "wasm"))]
+                    let waker = waker_fn(move || unsafe {
+                        let post_cobject: extern "C" fn(i64, *const core::ffi::c_void) =
+                            core::mem::transmute(_post_cobject);
+                        let obj: i32 = 0;
+                        post_cobject(port, &obj as *const _ as *const _);
```
Comment (wngr) [inline src/rust.rs:119]: What's `_post_cobject`?

File: tests/object.rs [hunk 2]
```diff
@@ -98,3 +98,90 @@ compile_pass! {
...
+    (
+        pub async fn create(value: u32) -> u32 {
+            value
+        }
+    ),
+    (
+        let _fut = __create(42);
+        let _f = __create_future_poll;
+        let _d = __create_future_drop;
+    ),
```
Comment (wngr) [inline tests/object.rs:114]: What does that do?

File: tests/object.rs [hunk 3]
```diff
@@ -98,3 +98,90 @@ compile_pass! {
...
+    (
+        final fut = api.create(42);
+        assert(await fut == 42);
+    ),
+    (
+        const fut = api.create(42);
+        assert.equal(await fut, 42);
+    ),
```
Comment (wngr) [inline tests/object.rs:122]: Pretty cool :+1:

File: tests/object.rs (hunk 2 above)
Comment (dvc94ch) [reply tests/object.rs:114]: check that the symbols are defined

File: src/rust.rs (hunk 1 above)
Comment (dvc94ch) [reply src/rust.rs:119]: it's allows you to send a message to a dart `ReceivePort`. this is the only way to asynchronously notify dart (from an arbitrary thread)

Comment (dvc94ch) [reply src/rust.rs:119]: it's a pointer to a function in the dart c api

File: tests/object.rs (hunk 3 above)
Comment (dvc94ch) [reply tests/object.rs:122]: now with streams...

File: src/js.rs [hunk 4]
```diff
@@ -113,6 +114,8 @@ impl JsGenerator {
                 });
             }
 
+            const { ReadableStream } = require("node:stream/web");
```
Comment (wngr) [inline src/js.rs:117]: In the browser this is in the global scope, so this should probably be guarded.. something like
```suggestion
            const { ReadableStream } = (typeof window == "object" && { ReadableStream }) || require("node:stream/web");
```

File: tests/object.rs [hunk 5]
```diff
@@ -185,3 +185,48 @@ compile_pass! {
...
+
+        pub fn create(values: &[u32]) -> impl Stream<Item = u32> {
+            TestStream(values.into_iter().rev().copied().collect())
+        }
+    ),
+    (
+        let values = [42, 99];
+        let stream = __create(values.as_ptr() as _, values.len() as _);
+        let _poll = __create_stream_poll;
+        __create_stream_drop(0, stream);
```
Comment (wngr) [inline tests/object.rs:211]: Why can't you assert the returned values?

**Review by wngr — APPROVED** [review]
image

**Comment by wngr** [comment]
I will look into the failing js test.

File: src/js.rs (hunk 4 above)
Comment (dvc94ch) [reply src/js.rs:117]: will that satisfy webpack? I started adapting the hack, but then thought that there's no point unless I actually test it, so I left it.

**Pull request merged**